### PR TITLE
[TOREE-537] Upgrade Spark 3.3.2

### PR DESCRIPTION
**Pull request by pan3793**
This is part of https://github.com/apache/incubator-toree/pull/199

Co-authored-by: Neil Skinner <reqpro@requaos.com>
Co-authored-by: Cheng Pan <chengpan@apache.org>

This PR aims to upgrade the Spark version to 3.3.2

The status of maintained Spark versions are

- Spark 3.0 => EOL
- Spark 3.1 => EOL
- Spark 3.2 => EOL
- Apache Spark 3.3 => December, 2023.
- Apache Spark 3.4 => October, 2024
- Upcoming Apache Spark 3.5 => 18 months from the release

Verified by `make dev`

image

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -61,7 +62,7 @@ RUN apt-get update && \
     rm -rf /var/lib/apt/lists/*
 
 ENV SPARK_HOME /usr/local/spark
-ENV PYTHONPATH $SPARK_HOME/python:$SPARK_HOME/python/lib/py4j-0.10.9-src.zip
+ENV PYTHONPATH $SPARK_HOME/python:$SPARK_HOME/python/lib/py4j-0.10.9.5-src.zip
```
Comment (pan3793) [inline Dockerfile:65]: keep consistent with Spark 3.3.2 https://github.com/apache/spark/blob/v3.3.2/python/lib/py4j-0.10.9.5-src.zip

File: project/Dependencies.scala [hunk 2]
```diff
@@ -36,10 +36,10 @@ object Dependencies {
   val coursier = "io.get-coursier" %% "coursier" % coursierVersion // Apache v2
   val coursierCache = "io.get-coursier" %% "coursier-cache" % coursierVersion // Apache v2
 
-  val ivy = "org.apache.ivy" % "ivy" % "2.4.0" // Apache v2
+  val ivy = "org.apache.ivy" % "ivy" % "2.5.1" // Apache v2
 
-  // use the same jackson version in test than the one provided at runtime by Spark 3.0.0
-  val jacksonDatabind = "com.fasterxml.jackson.core" % "jackson-databind" % "2.10.0" // Apache v2
+  // use the same jackson version in test than the one provided at runtime by Spark 3.3.2
+  val jacksonDatabind = "com.fasterxml.jackson.core" % "jackson-databind" % "2.13.4.2" // Apache v2
```
Comment (pan3793) [inline project/Dependencies.scala:42]: dependencies updated to align with Spark 3.3.2

https://mvnrepository.com/artifact/org.apache.spark/spark-core_2.12/3.3.2

File: build.sbt [hunk 3]
```diff
@@ -58,8 +58,8 @@ ThisBuild / javacOptions ++= Seq(
   "-Xlint:-options",
   "-Xlint:-processing",
   "-Werror",       // Treat warnings as errors
-  "-source", "1.6",
-  "-target", "1.6"
+  "-source", "1.8",
+  "-target", "1.8"
```
Comment (pan3793) [inline build.sbt:62]: Actually, Spark 3.x always releases with -target 1.8

File: build.sbt [hunk 4]
```diff
@@ -22,11 +22,11 @@ ThisBuild / version := Properties.envOrElse("VERSION", "0.0.0-dev") +
   (if ((ThisBuild / isSnapshot ).value) "-SNAPSHOT" else "")
 ThisBuild / isSnapshot := Properties.envOrElse("IS_SNAPSHOT","true").toBoolean
 ThisBuild / organization := "org.apache.toree.kernel"
-ThisBuild / crossScalaVersions := Seq("2.12.15")  // https://github.com/scala/bug/issues/12475, for Spark 3.2.0
+ThisBuild / crossScalaVersions := Seq("2.12.15")
```
Comment (pan3793) [inline build.sbt:25]: obviously, we should keep consistent with Spark 3.3.2

**Comment by pan3793** [comment]
ping @lresende, could you review this PR when you have time?

BTW, this is an awesome project, please let me know what I can do for this project :)
I'm familiar with Spark and Scala(but sorry for my poor Python skill), and am learning how to implement a Jupyter kernel using Java/Scala.

**Comment by lresende** [comment]
@pan3793 Thanks for all the enhancements, it is very welcomed. 

for this pr in specific, let's wait to merge it when the #199 gets merged... which, btw, if you want to help fix the remaining failing issues, we would welcome the help.

**Comment by pan3793** [comment]
@lresende thanks for your reply.

I actually took a look at #199 first, and thought it composes of

- upgrade spark 3.3
- upgrade scala 2.13
- upgrade some dependencies which the current version does not support scala 2.13, for example, cursior

This PR could be treated as part of that. And as I commented before

> Spark 3.0 ~ 3.5 uses Scala 2.12 by default, it rarely uses a non-default Scala version in production, so I would like to upgrade the Spark first.

would it be good to upgrade Scala 2.13 at the current stage?

**Comment by pan3793** [comment]
I updated the PR description to add @requaos(the original author of https://github.com/apache/incubator-toree/pull/199) and me as the co-author, from the tech view, I lean toward breaking down the huge PR into dedicated ones, and keep using Scala 2.12 during Spark 3.x serial.

@lresende WDYT? Of course, if you think we should work on https://github.com/apache/incubator-toree/pull/199, I would close this one and try to fix that.

**Comment by lresende** [comment]
> @lresende WDYT? Of course, if you think we should work on #199, I would close this one and try to fix that.

It's ok, let me test this and will review/merge based on that.

**Comment by requaos** [comment]
If this change is passable, then we can refocus #199 to introduce Scala 2.13, as upgrading Spark was ancillary to that effort. Let me know, because as is these commits will directly conflict with my branch for the overlap from copying my changes over instead of some other merge or cherrypick strategy. I will happily split up the efforts in my branch behind #199.

**Comment by pan3793** [comment]
> ... as is these commits will directly conflict with my branch for the overlap from copying my changes over instead of some other merge or cherrypick strategy.

@requaos Yes, this change starts from scratch, because I can not achieve it by cherry-picking in a clean way. I respect your effort in #199 so I added you as co-author of this PR. If necessary, I can help with git rebase and conflict resolution, which requires write permission of your forked repo.

**Review by lresende — APPROVED** [review]

**Pull request merged**